File: DsAlgoInScala/src/scala/programs/recursion/AllIndecies.scala
package scala.programs.recursion

object AllIndecies {

  def AllIndex(arr:Array[Int],ele:Int,idx:Int,fsf:Int):Array[Int] = arr match {
    case Array() => arr
    case Array(a) => if(a==ele) Array[Int](idx) else Array()
    case _ =>
      var res: Array[Int] = null
      if(idx==arr.length-1) {
        res = Array[Int](fsf)
        return res
      }
      if(ele==arr(idx)){
        res=AllIndex(arr,ele,idx+1,fsf+1)
        res(fsf) = idx
      } else {
        res=AllIndex(arr,ele,idx+1,fsf)
      }
      res
  }
  def main(args: Array[String]): Unit = {
    val r = AllIndex(Array(5,2,5,7),5,0,0)
    println(r)
  }
}

File: src/pep/recursion/practice/AllIndecies.java
package pep.recursion.practice;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class AllIndecies {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];

        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        int x = Integer.parseInt(br.readLine());
        int[] iarr = allIndices(arr, x, 0, 0);

        if (iarr.length == 0) {
            System.out.println();
            return;
        }

        for (int i = 0; i < iarr.length; i++) {
            System.out.println(iarr[i]);
        }
    }

    public static int[] allIndices(int[] arr, int x, int idx, int fsf) {
        // write ur code here
        int[] indecies;
        if (idx == arr.length - 1) {
            indecies = new int[fsf];
            return indecies;
        }
        if (x == arr[idx]) {
            indecies = allIndices(arr, x, idx + 1, fsf + 1);
            indecies[fsf] = idx;
        } else {
            indecies = allIndices(arr, x, idx + 1, fsf);
        }
        return indecies;
    }
}